gt/main.c

```diff
@@ -244,13 +247,233 @@ print_ip_err_msg(struct gt_packet_headers *pkt_info)
+	 */
+	write_seqlock(&eth_cache->lock);
+	ether_addr_copy(&map->ha, &eth_cache->eth_hdr.d_addr);
+	eth_cache->stale = map->stale;
+	write_sequnlock(&eth_cache->lock);
+
+	*pcall_again = true;
+}
+
+/*
+ * Fill up the Ethernet cached header.
+ * Note that the destination MAC address should be filled up by LLS.
+ */
+static int
+gt_fill_up_ether_cache_locked(struct ether_cache *eth_cache,
+	uint16_t inner_ip_ver, void *ip_dst, struct gatekeeper_if *iface)
+{
+	int ret;
+	unsigned lcore_id = rte_lcore_id();
+
+	eth_cache->stale = true;
+	eth_cache->ip_addr.proto = inner_ip_ver;
+
+	if (inner_ip_ver == ETHER_TYPE_IPv4) {
+		rte_memcpy(&eth_cache->ip_addr.ip.v4,
+			ip_dst, sizeof(eth_cache->ip_addr.ip.v4));
+	} else {
+		rte_memcpy(&eth_cache->ip_addr.ip.v6,
+			ip_dst, sizeof(eth_cache->ip_addr.ip.v6));
+	}
+
+	eth_cache->eth_hdr.ether_type = rte_cpu_to_be_16(inner_ip_ver);
+	ether_addr_copy(&iface->eth_addr, &eth_cache->eth_hdr.s_addr);
+	rte_atomic32_set(&eth_cache->ref_cnt, 1);
+
+	if (inner_ip_ver == ETHER_TYPE_IPv4) {
+		ret = hold_arp(gt_arp_and_nd_req_cb,
+			eth_cache, ip_dst, lcore_id);
+	} else {
+		ret = hold_nd(gt_arp_and_nd_req_cb,
+			eth_cache, ip_dst, lcore_id);
+	}
+
+	if (ret < 0)
+		clear_ether_cache(eth_cache);
+
+	return ret;
+}
+
+static int
+drop_cache_entry_randomly(struct neighbor_hash_table *neigh, uint16_t ip_ver)
+{
+	int ret;
+	uint32_t entry_id = rte_rand() % neigh->tbl_size;
+	struct ether_cache *eth_cache;
+	uint32_t entry_start_idx = entry_id;
+
+	while (true) {
+		eth_cache = &neigh->cache_tbl[entry_id];
+		if (rte_atomic32_read(&eth_cache->ref_cnt) == 0) {
+			entry_id = (entry_id + 1) % neigh->tbl_size;
+			eth_cache = NULL;
+		} else
+			break;
+
+		if (entry_start_idx == entry_id)
+			break;
+	}
+
+	if (eth_cache == NULL)
+		return -1;
+
+	if (ip_ver == ETHER_TYPE_IPv4) {
+		ret = put_arp(&eth_cache->ip_addr.ip.v4, rte_lcore_id());
+		if (ret < 0)
+			return ret;
+
+		ret = rte_hash_del_key(neigh->hash_table,
+			&eth_cache->ip_addr.ip.v4);
+		if (ret < 0) {
+			RTE_LOG(CRIT, GATEKEEPER,
+				"gt: failed to delete an Ethernet cache entry from the IPv4 neighbor table at %s, we are not trying to recover from this failure!",
+				__func__);
+		}
+		return ret;
+	}
+
+	if (likely(ip_ver == ETHER_TYPE_IPv6)) {
+		ret = put_nd(&eth_cache->ip_addr.ip.v6, rte_lcore_id());
+		if (ret < 0)
+			return ret;
+
+		ret = rte_hash_del_key(neigh->hash_table,
+			&eth_cache->ip_addr.ip.v6);
+		if (ret < 0) {
+			RTE_LOG(CRIT, GATEKEEPER,
+				"gt: failed to delete an Ethernet cache entry from the IPv6 neighbor table at %s, we are not trying to recover from this failure!",
+				__func__);
+		}
+		return ret;
+	}
+
+	return -1;
+}
+
+static struct ether_cache *
+get_new_ether_cache(struct neighbor_hash_table *neigh)
+{
+	int i;
+	struct ether_cache *eth_cache = NULL;
+
+	for (i = 0; i < neigh->tbl_size; i++) {
+		if (rte_atomic32_read(&neigh->cache_tbl[i].ref_cnt) == 0) {
+			eth_cache = &neigh->cache_tbl[i];
```

Comment:
Style: we can drop variable `eth_cache` and just return here. This will turn the whole body of the `if` statement, including its braces, into a simple `return &neigh->cache_tbl[i];`, and the current return at the end of the function into `return NULL;`.

Reply:
Done.